src/main/java/org/prlprg/rds/RDSLogger.java

```diff
@@ -0,0 +1,116 @@
+package org.prlprg.rds;
+
+import java.util.*;
+import java.util.function.Supplier;
+import java.util.logging.*;
+import org.prlprg.AppConfig;
+
+final class RDSLogger {
+  private final StringBuilder output = new StringBuilder();
+  private final boolean shouldLog;
+  private int indentLevel;
+
+  /**
+   * @param name a description of the read or write, printed at the start of logging
+   */
+  RDSLogger(String name, AppConfig.RDSLogLevel logLevel) {
+    this.indentLevel = 0;
+    this.shouldLog = AppConfig.RDS_LOG_LEVEL.compareTo(logLevel) >= 0;
+    logString(() -> name);
+  }
+
+  /** Globally enables log output to the provided handler */
+  public static void addHandler(Handler handler) {
+    var logger = Logger.getLogger(RDSLogger.class.getName());
+    logger.addHandler(handler);
+    handler.setLevel(Level.FINE);
+  }
+
+  /**
+   * Logs a lazily-evaluated String with the current indent level with Level.FINE.
+   *
+   * @param msg a thunk supplying the String to be logged.
+   */
+  private void logString(Supplier<String> msg) {
+    if (shouldLog) {
+      int level = this.indentLevel;
+      String indent = String.join("", Collections.nCopies(level, "  "));
+      output.append(indent).append(msg.get()).append("\n");
+    }
+  }
+
+  /**
+   * Logs the reading or writing of a new SEXP by printing an identifier and increasing the
+   * indentation level by 1
+   *
+   * @param sexpIdent a String identifying the SEXP being logged
+   */
+  public void push(String sexpIdent) {
```

Comment:
If using log4j2, indentation level can be controlled: https://howtodoinjava.com/log4j2/useful-conversion-pattern-examples/ (see left justified log level) and https://logging.apache.org/log4j/2.x/manual/layouts.html